### PR TITLE
Update 6P for better conformance with  IETF RFC 8480

### PR DESCRIPTION
**Pull request by yatch**
This PR updates 6P implementation for better conformance with [IETF RFC 8480](https://tools.ietf.org/html/rfc8480). Yet, some of behaviors defined in RFC 8480 are not implemented on purpose. For these, I put comments explaining why, to provide code-readers with clearer ideas behind.

In addition, this PR introduces `error` handler to `sixtop_sf_t`, which is used to inform a corresponding scheduling function of scheduling consistency detected on receiving a request. This change requires existing scheduling functions to have that handler; at least putting `NULL` at the end of `sixtop_sf_t` definitions is necessary.

New behavioral changes are tested manly by `test-sixtop-sf-error-handler.c` and `test-sixp.c`.

Once this PR is merged, I will update [the wiki](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-6TiSCH-6top-sub-layer).

Thank @debug-ito for his inputs and preliminary reviews.

### CODE REVIEW

**Comment by yatch** [comment]
I'm going to implement MSF on top of this.

**Comment by yatch** [comment]
'coap-lwm2m' test failed for some reason: https://travis-ci.org/contiki-ng/contiki-ng/jobs/549757701

I don't think the changes in this PR caused the failure. I will re-run the test to see if the test failure persists.

File: os/net/mac/tsch/sixtop/sixp.c [hunk 1]
```diff
@@ -190,16 +232,35 @@ sixp_input(const uint8_t *buf, uint16_t len, const linkaddr_t *src_addr)
...
-        LOG_ERR("6P: sixp_input() fails to return an error response");
+      if(pkt.code.cmd != SIXP_PKT_CMD_CLEAR &&
+         ((pkt.seqno == 0 && sixp_trans_get_seqno(trans) != 0) ||
+          (pkt.seqno != 0 && sixp_trans_get_seqno(trans) == 0))) {
+        /*
+         * seems the peer had power-cycle; we're going to send back
+         * RC_ERR_SEQNUM. in this case, we don't want to allocate
+         * another transaction for this new request.
+         */
+        goto err_seqnum;
```
Comment (debug-ito) [inline os/net/mac/tsch/sixtop/sixp.c]: Jumping into another conditional branch looks tricky to me. I would refactor the procedure under `err_seqnum` label into a function, and call it at the two places.

File: os/net/mac/tsch/sixtop/sixp.c [hunk 2]
```diff
@@ -351,13 +416,16 @@ sixp_output(sixp_pkt_type_t type, sixp_pkt_code_t code, uint8_t sfid,
...
   if(nbr == NULL &&
-     type == SIXP_PKT_TYPE_RESPONSE && code.value == SIXP_PKT_RC_SUCCESS &&
-     ((cmd = sixp_trans_get_cmd(trans)) == SIXP_PKT_CMD_ADD ||
-      cmd == SIXP_PKT_CMD_DELETE) &&
+     type == SIXP_PKT_TYPE_RESPONSE &&
+     sixp_trans_get_cmd(trans) != SIXP_PKT_CMD_CLEAR &&
+     (code.value != SIXP_PKT_RC_ERR_VERSION ||
+      code.value != SIXP_PKT_RC_ERR_SFID) &&
      (nbr = sixp_nbr_alloc(dest_addr)) == NULL) {
     LOG_ERR("6P: sixp_output() fails because of no memory for another nbr\n");
```
Comment (debug-ito) [inline os/net/mac/tsch/sixtop/sixp.c]: The condition `(code.value != SIXP_PKT_RC_ERR_VERSION || code.value != SIXP_PKT_RC_ERR_SFID)` is equal to `!(code.value == SIXP_PKT_RC_ERR_VERSION && code.value == SIXP_PKT_RC_ERR_SFID)`, which is always true. It's a mistake, right? 

I guess the correct condition is:

```c
nbr == NULL &&
type == SIXP_PKT_TYPE_RESPONSE &&
sixp_trans_get_cmd(trans) != SIXP_PKT_CMD_CLEAR &&
code.value != SIXP_PKT_RC_ERR_VERSION &&
code.value != SIXP_PKT_RC_ERR_SFID &&
(nbr = sixp_nbr_alloc(dest_addr)) == NULL
```

File: os/net/mac/tsch/sixtop/sixp.c (hunk 1 above)
Comment (yatch) [reply os/net/mac/tsch/sixtop/sixp.c]: I take your comment.

File: os/net/mac/tsch/sixtop/sixp.c (hunk 2 above)
Comment (yatch) [reply os/net/mac/tsch/sixtop/sixp.c]: Great catch! Fixing it.

**Comment by yatch** [comment]
This version of 6P works flawlessly for MSF (https://github.com/contiki-ng/contiki-ng/pull/1000) 👍

**Comment by alexrayne** [comment]
> This version of 6P works flawlessly with MSF (#1000) +1

Can you provide me receipt - how to mix this PR and PR#1000 into my fork? wich into to where merge?

**Comment by yatch** [comment]
Updated the wiki saying it supports RFC 8480: https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-6TiSCH-6top-sub-layer

**Pull request merged**